### PR TITLE
Add support for Link-Time Optimization

### LINKED ISSUE #64: Add support for LTO

### PR DESCRIPTION
**Pull request by antoyo**
Fix #64.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -27,6 +27,10 @@ gccjit = { git = "https://github.com/antoyo/gccjit.rs" }
 # Local copy.
 # gccjit = { path = "../gccjit.rs" }
 
+object = { version = "0.30.1", default-features = false, features = [
+    "std",
+    "read",
```
Comment (bjorn3) [inline Cargo.toml:32]: You may want to use `read_core` + the individual file formats + `unaligned` instead to avoid a dependency on `wasmparser`.

Comment (antoyo) [reply Cargo.toml:32]: Are you sure? It looks like the `wasm` stuff is not enabled by `read`.

Comment (bjorn3) [reply Cargo.toml:32]: You're right. My bad.

File: test.sh [hunk 2]
```diff
@@ -166,15 +167,18 @@ function build_sysroot() {
...
 
     echo "[AOT] alloc_system"
     $RUSTC example/alloc_system.rs --crate-type lib --target "$TARGET_TRIPLE"
 
+    # FIXME: it seems gcc will run lto1 unless -fno-lto is added.
     echo "[AOT] alloc_example"
-    $RUSTC example/alloc_example.rs --crate-type bin --target $TARGET_TRIPLE
+    # TODO: -Clink-arg=-lgcc_s
+    # TODO: -Cdefault-linker-libraries also fixes the issue.
+    PATH="$PATH:/opt/gcc/bin/" $RUSTC -Clto=yes example/alloc_example.rs -v --crate-type bin --target $TARGET_TRIPLE
```
Comment (antoyo) [inline test.sh:181]: Only enable LTO in the CI job with the sysroot compiled in release mode.

File: src/lib.rs [hunk 3]
```diff
@@ -321,6 +345,8 @@ pub fn __rustc_codegen_backend() -> Box<dyn CodegenBackend> {
 
     Box::new(GccCodegenBackend {
         target_info,
+        should_combine_object_files: Arc::new(AtomicBool::new(false)),
+        temp_dir: None,
```
Comment (antoyo) [inline src/lib.rs:349]: Make sure I delete the temp directory at the correct moment.

File: Cargo.toml [hunk 4]
```diff
@@ -20,13 +20,18 @@ harness = false
 [features]
 default = ["master"]
 master = ["gccjit/master"]
+# TODO: add feature for LTO, given all the setup that is needed (having same versions of gcc, lto1, lto-wrapper, liblto_plugin.so)?
```
Comment (antoyo) [inline Cargo.toml:23]: Don't forget this.

File: build_sysroot/Cargo.toml [hunk 5]
```diff
@@ -19,3 +19,5 @@ rustc-std-workspace-std = { path = "./sysroot_src/library/rustc-std-workspace-st
 
 [profile.release]
 debug = true
+# FIXME: is this actually needed?
```
Comment (antoyo) [inline build_sysroot/Cargo.toml:22]: Same.

File: src/back/lto.rs [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+        // bitcode. All modules were codegened in their own LLVM context, however,
+        // and we want to move everything to the same LLVM context. Currently the
+        // way we know of to do that is to serialize them to a string and them parse
+        // them later. Not great but hey, that's why it's "fat" LTO, right?
+        for module in in_memory {
+            // TODO: add the proper file extension here?
+            let path = tmp_path.path().to_path_buf().join(&module.name);
+            let path = path.to_str().expect("path");
+            let context = &module.module_llvm.context;
+            // TODO: do we need the optimization level here?
```
Comment (antoyo) [inline src/back/lto.rs:278]: Same.

File: test.sh [hunk 7]
```diff
@@ -198,8 +203,9 @@ function std_tests() {
     $RUSTC example/track-caller-attribute.rs --crate-type bin $TEST_FLAGS --target $TARGET_TRIPLE
     $RUN_WRAPPER ./target/out/track-caller-attribute
 
-    echo "[BUILD] mod_bench"
-    $RUSTC example/mod_bench.rs --crate-type bin --target $TARGET_TRIPLE
+    # FIXME: fail with sysroot built using LTO.
```
Comment (antoyo) [inline test.sh:206]: Try again.

File: Readme.md [hunk 8]
```diff
@@ -182,6 +182,10 @@ set substitute-path /usr/src/debug/gcc /path/to/gcc-repo/gcc
 
 TODO(antoyo): but that's not what I remember I was doing.
 
+### How to debug GCC LTO
+
+Run do the command with `-v -save-temps` and then extract the `lto1` line from the output and run that under the debugger
```
[the remark this answers is not in the document]
Comment (antoyo) [reply Readme.md]: I don't remember what's the output of GCC's `-save-temps`. Does cg_llvm output LLVM-specific stuff (except some stuff done manually like the bitcode used for LTO)?

[the remark this answers is not in the document]
Comment (bjorn3) [reply Readme.md]: LLVM doesn't produce intermediate files afaik. It keeps everything in memory until rustc explicitly tells it to either write bitcode or an object files.

[the remark this answers is not in the document]
Comment (antoyo) [reply Readme.md]: Ok, so I'll keep it that way for now and we'll see later.

**Pull request merged**